vscode/src/extension.ts

```diff
@@ -16,48 +16,138 @@ import { createSerialProvision } from "./toitProvision";
+  let message = "Could not find the `toit` executable";
+  if (path !== null) {
+    message += " at '" + path + "'";
+  }
+  message += ". Please make sure `toit` is installed and set the toit.path setting to the executable (reload the window to activate the extension";
+  const action = await Window.showErrorMessage(message, installAction, settingsAction);
+  if (action === installAction) {
+    env.openExternal(Uri.parse("https://docs.toit.io/getstarted/installation"));
+  } else if (action === settingsAction) {
+    Commands.executeCommand( "workbench.action.openSettings", "toit.path" );
+  }
+}
+
+async function badCLIPrompt(path: string|null) {
+  const installAction = "Install";
+  const settingsAction = "Update settings";
+  let message = "The `toit` executable ";
+  if (path !== null) {
+    message += "at '" + path + "' ";
+  }
+  message += "was found but did not execute correctly.";
+  const action = await Window.showErrorMessage(message, installAction, settingsAction);
   if (action === installAction) {
     env.openExternal(Uri.parse("https://docs.toit.io/getstarted/installation"));
   } else if (action === settingsAction) {
-    Commands.executeCommand( "workbench.action.openSettings", "toit.Path" );
+    Commands.executeCommand( "workbench.action.openSettings", "toit.path" );
   }
 }
 
+interface Executables {
+  cli: string | null;
+  lspCommand: Array<string> | null;
+}
+
+async function findExecutables(): Promise<Executables> {
+  const configCli = Workspace.getConfiguration("toit").get("path");
+  const configLspCommand = Workspace.getConfiguration("toitLanguageServer").get("command");
+
+  let cliExec: string|null = null;
+  let lspCommand: Array<string>|null = null;
+  let cliVersion: string|null = null;
+
+  if (configCli === null && configLspCommand === null) {
+    // No configuration. We try to find it in the PATH.
+    const check = await run("toit", [ "version", "-o", "short" ]);
+    if (check.executableExists) {
+      cliExec = "toit";
+      cliVersion = check.output;
+    } else {
+      // Try to find the language server in the path.
+      const lspResult = await run("toitlsp", ["version"]);
+      if (lspResult.executableExists && lspResult.output !== null) {
+        const toitcResult = await run("toitc", ["--version"]);
+        if (toitcResult.executableExists && toitcResult.output !== null) {
+          // Good enough for us.
+          lspCommand = [ "toitlsp", "--toitc", "toitc" ];
+        }
+      }
+    }
+  } else if (typeof configCli === "string") {
+    const check = await run(configCli, [ "version", "-o", "short" ]);
+    cliVersion = check.output;
+    if (check.executableExists) {
+      cliExec = configCli;
+      cliVersion = check.output;
+    }
+  }
+  if (configLspCommand !== null) {
+    lspCommand = configLspCommand as Array<string>;
+  }
+
+  if (cliExec === null && lspCommand === null) {
+    await missingCLIPrompt(configCli as string);
```

Comment:
How are we sure that `configCli` is a string at this point, and why is it necessary to have the `as`? Afterall `missingCLIPrompt` accepts null or string.

Reply:
Added checks and error message.